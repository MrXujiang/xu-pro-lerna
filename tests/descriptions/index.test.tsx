import { mount } from 'enzyme';
import React, { useRef } from 'react';
import { Button } from 'antd';
import ProDescriptions from '@ant-design/pro-descriptions';
import type { ProCoreActionType } from '@ant-design/pro-utils';
import { act } from 'react-dom/test-utils';

import { waitForComponentToPaint, waitTime } from '../util';

describe('descriptions', () => {
  it('🥩  descriptions render valueEnum when data = 0', async () => {
    const html = mount(
      <ProDescriptions
        columns={[
          {
            dataIndex: 'status',
            title: '状态',
            valueEnum: {
              0: { text: '关闭', status: 'Default' },
              1: { text: '运行中', status: 'Processing' },
              2: { text: '已上线', status: 'Success' },
              3: { text: '异常', status: 'Error' },
            },
          },
        ]}
        request={async () => ({
          data: {
            status: 0,
          },
        })}
      />,
    );
    await waitForComponentToPaint(html, 200);
    expect(html.find('span.ant-badge-status-text').text()).toBe('关闭');
  });

  it('🎏 onLoadingChange test', async () => {
    const fn = jest.fn();
    const html = mount(
      <ProDescriptions
        size="small"
        onLoadingChange={fn}
        columns={[
          {
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        request={async () => {
          return {
            data: [],
          };
        }}
      />,
    );
    await waitForComponentToPaint(html, 1200);
    expect(fn).toBeCalled();
  });

  it('🎏 loading test', async () => {
    const html = mount(
      <ProDescriptions
        columns={[
          {
            title: 'money',
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        request={async () => {
          return new Promise((resolve) => {
            setTimeout(() => {
              resolve({ data: [] });
            }, 5000);
          });
        }}
      />,
    );
    await waitForComponentToPaint(html, 1200);
    expect(html.find('.ant-skeleton').exists()).toBeTruthy();

    act(() => {
      html.setProps({
        loading: false,
      });
    });
    await waitForComponentToPaint(html, 1200);
    // props 指定为 false 后，无论 request 完成与否都不会出现 spin
    expect(html.find('.ant-skeleton').exists()).toBeFalsy();
  });

  it('🥩 test reload', async () => {
    const fn = jest.fn();
    const Reload = () => {
      const actionRef = useRef<ProCoreActionType>();
      return (
        <ProDescriptions
          actionRef={actionRef}
          title="高级定义列表 request"
          request={async () => {
            fn();
            await waitTime(200);
            return Promise.resolve({
              success: true,
              data: { id: '这是一段文本', date: '20200730', money: '12121' },
            });
          }}
          extra={
            <Button
              type="link"
              id="reload"
              onClick={() => {
                actionRef.current?.reload();
              }}
            >
              修改
            </Button>
          }
        >
          <ProDescriptions.Item label="文本" dataIndex="id" />
          <ProDescriptions.Item dataIndex="date" label="日期" valueType="date" />
          <ProDescriptions.Item label="money" dataIndex="money" valueType="money" />
        </ProDescriptions>
      );
    };
    const html = mount(<Reload />);
    await waitForComponentToPaint(html, 300);

    act(() => {
      html.find('Button#reload').simulate('click');
    });
    act(() => {
      html.find('Button#reload').simulate('click');
    });
    await waitForComponentToPaint(html);

    // 因为有 loading 的控制，所有只会触发两次
    expect(fn).toBeCalledTimes(2);
  });

  it('🥩 test reload by params', async () => {
    const fn = jest.fn();

    const html = mount(
      <ProDescriptions
        title="高级定义列表 request"
        request={async () => {
          fn();
          return Promise.resolve({
            success: true,
            data: { id: '这是一段文本', date: '20200730', money: '12121' },
          });
        }}
        extra={
          <Button type="link" id="reload">
            修改
          </Button>
        }
      >
        <ProDescriptions.Item label="文本" dataIndex="id" />
        <ProDescriptions.Item dataIndex="date" label="日期" valueType="date" />
        <ProDescriptions.Item label="money" dataIndex="money" valueType="money" />
      </ProDescriptions>,
    );
    await waitForComponentToPaint(html, 300);

    act(() => {
      html.setProps({
        params: { name: 'qixian' },
      });
    });

    await waitForComponentToPaint(html);

    expect(fn).toBeCalledTimes(2);
  });

  it('🥩 test request error', async () => {
    const fn = jest.fn();

    const html = mount(
      <ProDescriptions
        title="高级定义列表 request"
        request={async () => {
          throw new Error('load error');
        }}
        onRequestError={fn}
        extra={
          <Button type="link" id="reload">
            修改
          </Button>
        }
      >
        <ProDescriptions.Item label="文本" dataIndex="id" />
        <ProDescriptions.Item dataIndex="date" label="日期" valueType="date" />
        <ProDescriptions.Item label="money" dataIndex="money" valueType="money" />
      </ProDescriptions>,
    );
    await waitForComponentToPaint(html, 300);

    expect(fn).toBeCalledTimes(1);
  });

  it('🏊‍♂️ Progress', () => {
    const html = mount(
      <ProDescriptions>
        <ProDescriptions.Item label="进度条1" valueType="progress">
          40
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条2" valueType="progress">
          -1
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条3" valueType="progress">
          100
        </ProDescriptions.Item>
      </ProDescriptions>,
    );
    expect(html.find('.ant-progress-text').at(0).text()).toEqual('40%');
    expect(html.find('.ant-progress-text').at(1).find('.anticon-close-circle')).toBeTruthy();
    expect(html.find('.ant-progress-text').at(1).find('.anticon-check-circle')).toBeTruthy();
  });

  it('🏊‍♂️ ProDescriptions support order', () => {
    const html = mount(
      <ProDescriptions
        dataSource={{
          title: 'test',
        }}
        columns={[
          {
            title: '标题',
            dataIndex: 'title',
            valueType: 'text',
            order: 100,
          },
        ]}
      >
        <ProDescriptions.Item order={9} label="进度条1" valueType="progress">
          40
        </ProDescriptions.Item>
        <ProDescriptions.Item label="进度条2" valueType="progress">
          -1
        </ProDescriptions.Item>
        <ProDescriptions.Item order={8} label="进度条3" valueType="progress">
          100
        </ProDescriptions.Item>
      </ProDescriptions>,
    );
    act(() => {
      expect(html.render()).toMatchSnapshot();
    });
  });
});
