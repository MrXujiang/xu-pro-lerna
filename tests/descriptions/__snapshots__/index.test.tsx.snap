// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`descriptions üèä‚Äç‚ôÇÔ∏è ProDescriptions support order 1`] = `
<div
  class="ant-descriptions ant-pro-descriptions"
>
  <div
    class="ant-descriptions-view"
  >
    <table>
      <tbody>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                Ê†áÈ¢ò
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                test
              </span>
            </div>
          </td>
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                ËøõÂ∫¶Êù°1
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                <div
                  class="ant-progress ant-progress-line ant-progress-status-active ant-progress-show-info ant-progress-small"
                  style="min-width: 100px; max-width: 320px;"
                >
                  <div
                    class="ant-progress-outer"
                  >
                    <div
                      class="ant-progress-inner"
                    >
                      <div
                        class="ant-progress-bg"
                        style="width: 40%; height: 6px;"
                      />
                    </div>
                  </div>
                  <span
                    class="ant-progress-text"
                    title="40%"
                  >
                    40%
                  </span>
                </div>
              </span>
            </div>
          </td>
          <td
            class="ant-descriptions-item"
            colspan="1"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                ËøõÂ∫¶Êù°3
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                <div
                  class="ant-progress ant-progress-line ant-progress-status-success ant-progress-show-info ant-progress-small"
                  style="min-width: 100px; max-width: 320px;"
                >
                  <div
                    class="ant-progress-outer"
                  >
                    <div
                      class="ant-progress-inner"
                    >
                      <div
                        class="ant-progress-bg"
                        style="width: 100%; height: 6px;"
                      />
                    </div>
                  </div>
                  <span
                    class="ant-progress-text"
                  >
                    <span
                      aria-label="check-circle"
                      class="anticon anticon-check-circle"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="check-circle"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </span>
            </div>
          </td>
        </tr>
        <tr
          class="ant-descriptions-row"
        >
          <td
            class="ant-descriptions-item"
            colspan="3"
          >
            <div
              class="ant-descriptions-item-container"
            >
              <span
                class="ant-descriptions-item-label"
              >
                ËøõÂ∫¶Êù°2
              </span>
              <span
                class="ant-descriptions-item-content"
              >
                <div
                  class="ant-progress ant-progress-line ant-progress-status-exception ant-progress-show-info ant-progress-small"
                  style="min-width: 100px; max-width: 320px;"
                >
                  <div
                    class="ant-progress-outer"
                  >
                    <div
                      class="ant-progress-inner"
                    >
                      <div
                        class="ant-progress-bg"
                        style="width: 0%; height: 6px;"
                      />
                    </div>
                  </div>
                  <span
                    class="ant-progress-text"
                  >
                    <span
                      aria-label="close-circle"
                      class="anticon anticon-close-circle"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="close-circle"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
